difftest/t1-simulator/src/difftest.rs

```diff
@@ -0,0 +1,203 @@
+mod dut;
+mod spike;
+
+use dut::*;
+pub use spike::SpikeHandle;
+use std::path::Path;
+use tracing::{error, trace};
+
+pub struct Difftest {
+  spike: SpikeHandle,
+  dut: Dut,
+}
+
+impl Difftest {
+  pub fn new(
+    size: usize,
+    elf_file: String,
+    log_file: String,
+    vlen: u32,
+    dlen: u32,
+    set: String,
+  ) -> Self {
+    Self {
+      spike: SpikeHandle::new(size, Path::new(&elf_file), vlen, dlen, set),
+      dut: Dut::new(Path::new(&log_file)),
+    }
+  }
+
+  fn peek_issue(&mut self, issue: IssueEvent) -> anyhow::Result<()> {
+    self.spike.peek_issue(issue).unwrap();
+
+    Ok(())
+  }
+
+  fn update_lsu_idx(&mut self, lsu_enq: LsuEnqEvent) -> anyhow::Result<()> {
+    self.spike.update_lsu_idx(lsu_enq).unwrap();
+
+    Ok(())
+  }
+
+  fn poke_inst(&mut self) -> anyhow::Result<()> {
+    loop {
+      let se = self.spike.find_se_to_issue();
+      if (se.is_vfence_insn || se.is_exit_insn) && self.spike.to_rtl_queue.len() == 1 {
+        if se.is_exit_insn {
+          error!("Simulation quit graceful");
+          return Err(anyhow::anyhow!("graceful exit"));
+        }
+
+        self.spike.to_rtl_queue.pop_back();
+      } else {
+        break;
+      }
+    }
+
+    // TODO: remove these, now just for aligning online difftest
+    if let Some(se) = self.spike.to_rtl_queue.front() {
+      // it is ensured there are some other instruction not committed, thus
+      // se_to_issue should not be issued
+      if se.is_vfence_insn || se.is_exit_insn {
+        assert!(
+          self.spike.to_rtl_queue.len() > 1,
+          "to_rtl_queue are smaller than expected"
+        );
+        if se.is_exit_insn {
+          trace!("DPIPokeInst: exit waiting for fence");
+        } else {
+          trace!("DPIPokeInst: waiting for fence, no issuing new instruction");
+        }
+      } else {
+        trace!(
+          "DPIPokeInst: poke instruction: pc={:#x}, inst={}",
+          se.pc,
+          se.disasm
+        );
+      }
+    }
+    Ok(())
+  }
+
+  pub fn diff(&mut self) -> anyhow::Result<()> {
+    self.poke_inst().unwrap();
+
+    let event = self.dut.step()?;
+
+    match &*event.event {
+      "peekTL" => {
```

Comment:
In the following PR, please rename it to "memoryWrite", in the next architecture change, we will refactor T1 to AXI.